TechnicalDocuments/Proposals/Fish_Operator.md

```diff
@@ -0,0 +1,124 @@
+# Introduce "fish" operator
+
+* Author: Ilya Puchka
+* Review Manager:TBD
+
+## Introduction
+
+This proposal it to suggest to use new operator that will allow us to compose functions returning optional values, which is not currently possible.
+
+## Motivation
+
+Since introduction of `>>>` operator it was widely adopted throughout the code base and became an important tool in our tool set. But there are some cases where it's not enough. 
+
+One of these cases is when first item returns optional, which makes us to jump through the hoops:
+
+```swift
+didTap: helpText.map { ^Action.didTapWhatDoesThisMean($0) >>> observer }
+```
+
+## Proposed solution
+
+To handle such case [Prelude](https://github.com/pointfreeco/swift-prelude) defines "fish" operator (with a help of free curried `flatMap` function):
+
+```swift
+infix operator >=>: ForwardComposition
+
+public func >=> <A, B, C>(lhs: @escaping (A) -> B?, rhs: @escaping (B) -> C?) -> (A) -> C? {
+  return lhs >>> flatMap(rhs)
+}
+
+public func flatMap<A, B>(_ a2b: @escaping (A) -> B?) -> (A?) -> B? {
+  return { a in
+    a.flatMap(a2b)
+  }
+}
+```
+
+With this we can convert example code above to much simpler form:
+
+```swift
+didTap: ^helpText >=> Action.didTapWhatDoesThisMean >=> observer
+```
+
+When we will look into types we will see the following:
+
+```swift
+() -> String? >=> (String) -> Action >=> (Action) -> Void
+```
+
+This will not compile though as Swift treats `() -> A` and `(()) -> A` as different types. And even if it would treat them the same the final function that would be created by this chain will have signature `() -> Void?` which is not the same as `() -> Void`.
+
+To solve that we can introduce few overload of `>=>` operator to handle these special cases:
+
+```swift
+public func >=> <A, B>(lhs: @escaping () -> A?, rhs: @escaping (A) -> B?) -> () -> B? {
+    return lhs >>> flatMap(rhs)
+}
+
+public func >=> <A, B>(lhs: @escaping (A) -> B?, rhs: @escaping (B) -> Void) -> (A) -> Void {
+    return lhs >>> flatMap(rhs)
+}
+
+public func >=> <A>(lhs: @escaping () -> A?, rhs: @escaping (A) -> Void) -> () -> Void {
+    return lhs >>> flatMap(rhs)
+}
+
+public func flatMap<A>(_ a2b: @escaping (A) -> Void) -> (A?) -> Void {
+    return { a in
+        a.flatMap(a2b)
+    }
+}
+```
+
+This can result in ambiguty in some cases and will force to specify types explicitly, so to avoid that we can alternatively introduce new operator to work specifically with `Void` types:
```

Comment:
So, is this the proposal? To introduce the `>->` operator only? Does this operator conflict with anything commonly used?

Reply:
This is alternative solution to avoid possible ambiguities (though I think such cases will be rare so it is not needed and can be addressed when we face this issue.) I'll move this to "alternatives" section.

Reply:
I mean the proposal is to introduce `>=>` as a single operator with overloads for `Void` types, but alternatively we can introduce both `>=>` and `>->` (last just for `Void` types).

Reply:
I was also confused with that was being proposed, my suggestion is to keep this proposal focused on fish operator and not introduce anything else, the rest should go into its own proposal for discussion.

Reply:
Removed these details and moved this to the alternatives section